Many unimplemented storage paths in WebsiteDataStore.cpp

mcatanzaro opened the pull request:
#### 4993fe63cf51c7b4631445966440aa08fbf27e32
<pre>
Many unimplemented storage paths in WebsiteDataStore.cpp
<a href="https://bugs.webkit.org/show_bug.cgi?id=251808">https://bugs.webkit.org/show_bug.cgi?id=251808</a>

Reviewed by Sihui Liu.

Currently we define default storage directories in platform-specific
source files. Originally this probably seemed like a good idea to avoid
cluttering the cross-platform file with platform-specific details.
However, for non-Cocoa ports, there is nowadays no difference in
implementation except for the names selected for the directories, which
are different for no good reason but which we cannot easily change. All
the platform-specific logic has already moved to the functions
WebsiteDataStore::cacheDirectoryFileSystemRepresentation and
WebsiteDataStore::websiteDataDirectoryFileSystemRepresentation. So let&apos;s
move all the default directory names to the cross-platform file, except
for Cocoa ports, where the implementations are still slightly different.

The primary goal of this change is to encourage developers to implement
the functions when adding them. Since the implementation is as simple as
choosing whether to use cache or data directory, plus choosing the name
for the directory, there&apos;s really no good reason not to. Otherwise, it
could take seven years for us to come in and fill in a directory name,
as happened for defaultMediaCacheDirectory. Developers probably
occasionally look in platform-specific files to look for unimplemented
functions, but not in cross-platform files, so having unimplemented
platform-specific code in cross-platform files is particularly annoying.

For the most part, I selected the Windows port names to be the new
default directory name, since the Windows names match the Cocoa names
(except for the media keys directory) and look nicer. PlayStation and
GLib port names are treated as exceptions. Hopefully this will serve as
a hint to future developers to treat them accordingly and not copy them.
The names of future directory types should match the Windows and Cocoa
names, not the GLib and PlayStation names, since there&apos;s no good reason
for them to be different.

This commits makes a few additional changes. First, it defines
HSTSStorageDirectory always. It&apos;s OK for it to be unused on certain
ports. Next, it implements defaultMediaCacheDirectory (implementation
missing since 2016!) and defaultAlternativeServicesDirectory (which is
unused, but there&apos;s no reason not to define where it would go if it were
not unused). It leaves defaultJavaScriptConfigurationDirectory
unimplemented because this is the only config directory type we have and
it really does not belong under data or cache, and this data type is
also unused except on Cocoa ports. I believe the only behavior change
here should be the addition of the media cache directory.

A GLib-specific bug in the defaultHSTSStorageDirectory is fixed, but
only for the new API versions. The HSTS database goes under the base
cache directory if provided, but if not provided it goes under the base
data directory. Oops. Leave it broken in the original API versions to
avoid leaking the HSTS database on disk. Also, move it to a subdirectory
in the new API version.

Finally, this commit moves WebsiteDataStoreGLib.cpp itself to live with
the other WebsiteData files. It was a bit lost previously.

* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/SourcesWPE.txt:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::defaultCacheStorageDirectory):
(WebKit::WebsiteDataStore::defaultGeneralStorageDirectory):
(WebKit::WebsiteDataStore::defaultNetworkCacheDirectory):
(WebKit::WebsiteDataStore::defaultApplicationCacheDirectory):
(WebKit::WebsiteDataStore::defaultMediaCacheDirectory):
(WebKit::WebsiteDataStore::defaultIndexedDBDatabaseDirectory):
(WebKit::WebsiteDataStore::defaultServiceWorkerRegistrationDirectory):
(WebKit::WebsiteDataStore::defaultWebSQLDatabaseDirectory):
(WebKit::WebsiteDataStore::defaultHSTSStorageDirectory):
(WebKit::WebsiteDataStore::defaultLocalStorageDirectory):
(WebKit::WebsiteDataStore::defaultMediaKeysStorageDirectory):
(WebKit::WebsiteDataStore::defaultAlternativeServicesDirectory):
(WebKit::WebsiteDataStore::defaultDeviceIdHashSaltsStorageDirectory):
(WebKit::WebsiteDataStore::defaultResourceLoadStatisticsDirectory):
(WebKit::WebsiteDataStore::defaultJavaScriptConfigurationDirectory):
(WebKit::WebsiteDataStore::defaultUnifiedOriginStorageLevel): Deleted.
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.cpp:
(WebKit::WebsiteDataStoreConfiguration::initializePaths):
* Source/WebKit/UIProcess/WebsiteData/glib/WebsiteDataStoreGLib.cpp: Renamed from Source/WebKit/UIProcess/glib/WebsiteDataStoreGLib.cpp.
(WebKit::programName):
(WebKit::WebsiteDataStore::defaultBaseCacheDirectory):
(WebKit::WebsiteDataStore::defaultBaseDataDirectory):
(WebKit::WebsiteDataStore::cacheDirectoryFileSystemRepresentation):
(WebKit::WebsiteDataStore::websiteDataDirectoryFileSystemRepresentation):
(WebKit::WebsiteDataStore::defaultUnifiedOriginStorageLevel):
* Source/WebKit/UIProcess/WebsiteData/playstation/WebsiteDataStorePlayStation.cpp:
(WebKit::WebsiteDataStore::defaultUnifiedOriginStorageLevel):
(WebKit::WebsiteDataStore::defaultApplicationCacheDirectory): Deleted.
(WebKit::WebsiteDataStore::defaultCacheStorageDirectory): Deleted.
(WebKit::WebsiteDataStore::defaultGeneralStorageDirectory): Deleted.
(WebKit::WebsiteDataStore::defaultNetworkCacheDirectory): Deleted.
(WebKit::WebsiteDataStore::defaultIndexedDBDatabaseDirectory): Deleted.
(WebKit::WebsiteDataStore::defaultServiceWorkerRegistrationDirectory): Deleted.
(WebKit::WebsiteDataStore::defaultLocalStorageDirectory): Deleted.
(WebKit::WebsiteDataStore::defaultMediaKeysStorageDirectory): Deleted.
(WebKit::WebsiteDataStore::defaultWebSQLDatabaseDirectory): Deleted.
(WebKit::WebsiteDataStore::defaultResourceLoadStatisticsDirectory): Deleted.
* Source/WebKit/UIProcess/WebsiteData/win/WebsiteDataStoreWin.cpp:
(WebKit::WebsiteDataStore::defaultUnifiedOriginStorageLevel):
(WebKit::WebsiteDataStore::defaultApplicationCacheDirectory): Deleted.
(WebKit::WebsiteDataStore::defaultCacheStorageDirectory): Deleted.
(WebKit::WebsiteDataStore::defaultNetworkCacheDirectory): Deleted.
(WebKit::WebsiteDataStore::defaultGeneralStorageDirectory): Deleted.
(WebKit::WebsiteDataStore::defaultIndexedDBDatabaseDirectory): Deleted.
(WebKit::WebsiteDataStore::defaultServiceWorkerRegistrationDirectory): Deleted.
(WebKit::WebsiteDataStore::defaultLocalStorageDirectory): Deleted.
(WebKit::WebsiteDataStore::defaultMediaKeysStorageDirectory): Deleted.
(WebKit::WebsiteDataStore::defaultWebSQLDatabaseDirectory): Deleted.
(WebKit::WebsiteDataStore::defaultResourceLoadStatisticsDirectory): Deleted.

Canonical link: <a href="https://commits.webkit.org/260511@main">https://commits.webkit.org/260511@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4efbff3fe15987ae3cdf0dde069bb2047d849033

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/108507 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/17604 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/41357 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/117613 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/117816 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/19055 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/8882 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/100735 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114274 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/14286 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/97511 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/42246 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/96264 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/29151 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/83954 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/10414 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/30500 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/11171 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/7409 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/16561 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/50096 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7265 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/12759 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -1969,21 +1969,143 @@ void WebsiteDataStore::resetStoragePersistedState(CompletionHandler<void()>&& co
+// So confusingly, this should use baseDataDirectory rather than baseCacheDirectory. However, we
+// have no migrator to move or -- at minimum -- delete the data from the old location. We should at
+// least ensure deletion before moving the location.
+String WebsiteDataStore::defaultCacheStorageDirectory(const String& baseCacheDirectory)
```
I don't think this needs FIXME as this patch does not change current behavior, and I am not sure what needs to be fixed (looks like we are not going to change `cacheDirectoryFileSystemRepresentation` to `websiteDataDirectoryFileSystemRepresentation`). 

We could add a non-FIXME comment for the rationale of using cache directory, like: "CacheStorage should be under data directory per spec https://w3c.github.io/ServiceWorker/#cache-lifetimes. Keep using cache directory as legacy implementation because we will soon use GeneralStorageDirectory for CacheStorage (webkit.org/b/244058) and migrate existing data to it."

mcatanzaro replied:
OK.

mcatanzaro replied:
I've rewritten this comment. Please review again to make sure it's correct.

szewai commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -1969,21 +1969,143 @@ void WebsiteDataStore::resetStoragePersistedState(CompletionHandler<void()>&& co
+String WebsiteDataStore::defaultHSTSStorageDirectory(const String& baseCacheDirectory)
+{
+#if USE(GLIB)
+    return websiteDataDirectoryFileSystemRepresentation(""_s, baseCacheDirectory);
```
I know this is existing behavior but it looks a bit dangerous, as I think we usually check if directory is empty to decide if we a type can be stored. So it seems this should either be websiteDataDirectoryFileSystemRepresentation("HSTS"_s, baseDataDirectory), if you want to create a HSTS storage, or emptyString() (unless you do want to store this under the root cache directory...)

And it seems weird that this calls `websiteDataDirectoryFileSystemRepresentation` with `baseCacheDirectory`...

mcatanzaro replied:
> I know this is existing behavior but it looks a bit dangerous, as I think we usually check if directory is empty to decide if we a type can be stored. So it seems this should either be websiteDataDirectoryFileSystemRepresentation("HSTS"_s, baseDataDirectory), if you want to create a HSTS storage, or emptyString() (unless you do want to store this under the root cache directory...)

The code is working properly: the HSTS database really does go in the root cache directory. This might look weird, but it's suitable for GLib because the HSTS database is exactly one file, and creating a directory to store only one file is kinda overkill.

Ideally the HSTS database would be partitioned by your OriginStorageManager to prevent HSTS supercookie abuse. Well, perhaps _all_ website data should be partitioned. But anyway, for now it's just one file.

> And it seems weird that this calls websiteDataDirectoryFileSystemRepresentation with baseCacheDirectory...

Oh good catch. That's a preexisting bug, dating all the way back to https://commits.webkit.org/214897@main. O_O But it only happens if the application does not configure a base cache directory, so it doesn't affect Epiphany or MiniBrowser. Probably does affect most other applications, though. Drat.

I'll have to think about this. I'll fix it for the new GLib API for sure. For the old API, if I fix it, then the old HSTS database at the old location will be leaked on disk. That doesn't seem great, but I don't really want to write a migrator, so maybe I will leave the old API version untouched. Hmm....

mcatanzaro replied:
I decided it's easiest to fix it only for the new API.

mcatanzaro replied:
BTW if you recommend moving it under a directory instead of putting it in the base cache directory, I suppose now is the best time to do that. I don't think it matters too much either way.

mcatanzaro replied:
> I know this is existing behavior but it looks a bit dangerous, as I think we usually check if directory is empty to decide if we a type can be stored. So it seems this should either be websiteDataDirectoryFileSystemRepresentation("HSTS"_s, baseDataDirectory), if you want to create a HSTS storage, or emptyString() (unless you do want to store this under the root cache directory...)

I decided to move it in the new API version.

szewai commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -1969,21 +1969,147 @@ void WebsiteDataStore::resetStoragePersistedState(CompletionHandler<void()>&& co
+// But we cannot change this function to use baseDataDirectory without leaking previous storage.
+// This problem will go away when using UnifiedOriginStorageLevel::Standard, as then data will be
+// migrated from this location to the GeneralStorageDirectory.
+String WebsiteDataStore::defaultCacheStorageDirectory(const String& baseCacheDirectory)
```
The comment should probably be put inside function as it explains why we use `baseCacheDirectory`. I prefer the comment to be more succinct like:

> This should be baseDataDirectory per spec: https://w3c.github.io/ServiceWorker/#cache-lifetimes.
> Keep using baseCacheDirectory now as we plan to use GeneralStorageDirectory for CacheStorage soon.

(I think we don't normally comment "we do this because someone said so", but "we do this because of some good reason" : P)

mcatanzaro replied:
OK, I reworded this to:

>     // CacheStorage is really data, not cache, as its lifetime should be controlled by web clients.
>     // https://w3c.github.io/ServiceWorker/#cache-lifetimes
>     //
>     // Keep using baseCacheDirectory for now for compatibility, to avoid leaking existing storage.
>     // Soon, it will be migrated to GeneralStorageDirectory.

szewai reviewed (approved):
r=me

mcatanzaro committed:
Many unimplemented storage paths in WebsiteDataStore.cpp
https://bugs.webkit.org/show_bug.cgi?id=251808

Reviewed by Sihui Liu.

Currently we define default storage directories in platform-specific
source files. Originally this probably seemed like a good idea to avoid
cluttering the cross-platform file with platform-specific details.
However, for non-Cocoa ports, there is nowadays no difference in
implementation except for the names selected for the directories, which
are different for no good reason but which we cannot easily change. All
the platform-specific logic has already moved to the functions
WebsiteDataStore::cacheDirectoryFileSystemRepresentation and
WebsiteDataStore::websiteDataDirectoryFileSystemRepresentation. So let's
move all the default directory names to the cross-platform file, except
for Cocoa ports, where the implementations are still slightly different.

The primary goal of this change is to encourage developers to implement
the functions when adding them. Since the implementation is as simple as
choosing whether to use cache or data directory, plus choosing the name
for the directory, there's really no good reason not to. Otherwise, it
could take seven years for us to come in and fill in a directory name,
as happened for defaultMediaCacheDirectory. Developers probably
occasionally look in platform-specific files to look for unimplemented
functions, but not in cross-platform files, so having unimplemented
platform-specific code in cross-platform files is particularly annoying.

For the most part, I selected the Windows port names to be the new
default directory name, since the Windows names match the Cocoa names
(except for the media keys directory) and look nicer. PlayStation and
GLib port names are treated as exceptions. Hopefully this will serve as
a hint to future developers to treat them accordingly and not copy them.
The names of future directory types should match the Windows and Cocoa
names, not the GLib and PlayStation names, since there's no good reason
for them to be different.

This commits makes a few additional changes. First, it defines
HSTSStorageDirectory always. It's OK for it to be unused on certain
ports. Next, it implements defaultMediaCacheDirectory (implementation
missing since 2016!) and defaultAlternativeServicesDirectory (which is
unused, but there's no reason not to define where it would go if it were
not unused). It leaves defaultJavaScriptConfigurationDirectory
unimplemented because this is the only config directory type we have and
it really does not belong under data or cache, and this data type is
also unused except on Cocoa ports. I believe the only behavior change
here should be the addition of the media cache directory.

A GLib-specific bug in the defaultHSTSStorageDirectory is fixed, but
only for the new API versions. The HSTS database goes under the base
cache directory if provided, but if not provided it goes under the base
data directory. Oops. Leave it broken in the original API versions to
avoid leaking the HSTS database on disk. Also, move it to a subdirectory
in the new API version.

Finally, this commit moves WebsiteDataStoreGLib.cpp itself to live with
the other WebsiteData files. It was a bit lost previously.

* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/SourcesWPE.txt:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::defaultCacheStorageDirectory):
(WebKit::WebsiteDataStore::defaultGeneralStorageDirectory):
(WebKit::WebsiteDataStore::defaultNetworkCacheDirectory):
(WebKit::WebsiteDataStore::defaultApplicationCacheDirectory):
(WebKit::WebsiteDataStore::defaultMediaCacheDirectory):
(WebKit::WebsiteDataStore::defaultIndexedDBDatabaseDirectory):
(WebKit::WebsiteDataStore::defaultServiceWorkerRegistrationDirectory):
(WebKit::WebsiteDataStore::defaultWebSQLDatabaseDirectory):
(WebKit::WebsiteDataStore::defaultHSTSStorageDirectory):
(WebKit::WebsiteDataStore::defaultLocalStorageDirectory):
(WebKit::WebsiteDataStore::defaultMediaKeysStorageDirectory):
(WebKit::WebsiteDataStore::defaultAlternativeServicesDirectory):
(WebKit::WebsiteDataStore::defaultDeviceIdHashSaltsStorageDirectory):
(WebKit::WebsiteDataStore::defaultResourceLoadStatisticsDirectory):
(WebKit::WebsiteDataStore::defaultJavaScriptConfigurationDirectory):
(WebKit::WebsiteDataStore::defaultUnifiedOriginStorageLevel): Deleted.
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.cpp:
(WebKit::WebsiteDataStoreConfiguration::initializePaths):
* Source/WebKit/UIProcess/WebsiteData/glib/WebsiteDataStoreGLib.cpp: Renamed from Source/WebKit/UIProcess/glib/WebsiteDataStoreGLib.cpp.
(WebKit::programName):
(WebKit::WebsiteDataStore::defaultBaseCacheDirectory):
(WebKit::WebsiteDataStore::defaultBaseDataDirectory):
(WebKit::WebsiteDataStore::cacheDirectoryFileSystemRepresentation):
(WebKit::WebsiteDataStore::websiteDataDirectoryFileSystemRepresentation):
(WebKit::WebsiteDataStore::defaultUnifiedOriginStorageLevel):
* Source/WebKit/UIProcess/WebsiteData/playstation/WebsiteDataStorePlayStation.cpp:
(WebKit::WebsiteDataStore::defaultUnifiedOriginStorageLevel):
(WebKit::WebsiteDataStore::defaultApplicationCacheDirectory): Deleted.
(WebKit::WebsiteDataStore::defaultCacheStorageDirectory): Deleted.
(WebKit::WebsiteDataStore::defaultGeneralStorageDirectory): Deleted.
(WebKit::WebsiteDataStore::defaultNetworkCacheDirectory): Deleted.
(WebKit::WebsiteDataStore::defaultIndexedDBDatabaseDirectory): Deleted.
(WebKit::WebsiteDataStore::defaultServiceWorkerRegistrationDirectory): Deleted.
(WebKit::WebsiteDataStore::defaultLocalStorageDirectory): Deleted.
(WebKit::WebsiteDataStore::defaultMediaKeysStorageDirectory): Deleted.
(WebKit::WebsiteDataStore::defaultWebSQLDatabaseDirectory): Deleted.
(WebKit::WebsiteDataStore::defaultResourceLoadStatisticsDirectory): Deleted.
* Source/WebKit/UIProcess/WebsiteData/win/WebsiteDataStoreWin.cpp:
(WebKit::WebsiteDataStore::defaultUnifiedOriginStorageLevel):
(WebKit::WebsiteDataStore::defaultApplicationCacheDirectory): Deleted.
(WebKit::WebsiteDataStore::defaultCacheStorageDirectory): Deleted.
(WebKit::WebsiteDataStore::defaultNetworkCacheDirectory): Deleted.
(WebKit::WebsiteDataStore::defaultGeneralStorageDirectory): Deleted.
(WebKit::WebsiteDataStore::defaultIndexedDBDatabaseDirectory): Deleted.
(WebKit::WebsiteDataStore::defaultServiceWorkerRegistrationDirectory): Deleted.
(WebKit::WebsiteDataStore::defaultLocalStorageDirectory): Deleted.
(WebKit::WebsiteDataStore::defaultMediaKeysStorageDirectory): Deleted.
(WebKit::WebsiteDataStore::defaultWebSQLDatabaseDirectory): Deleted.
(WebKit::WebsiteDataStore::defaultResourceLoadStatisticsDirectory): Deleted.

Canonical link: https://commits.webkit.org/260511@main
